x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,84 +358,172 @@ public void removeFromCache(final CacheKey cacheKey) {
     /**
      * Marks the specified searchable snapshot shard as evicted in cache. Cache files associated with this shard will be evicted from cache.
      *
-     * @param snapshotId the {@link SnapshotId}
-     * @param indexId the {@link SnapshotId}
-     * @param shardId the {@link SnapshotId}
+     * @param snapshotUUID      the snapshot's unique identifier
+     * @param snapshotIndexName the name of the index in the snapshot
+     * @param shardId           the {@link ShardId}
      */
-    public void markShardAsEvictedInCache(SnapshotId snapshotId, IndexId indexId, ShardId shardId) {
-        final ShardEviction shardEviction = new ShardEviction(snapshotId.getUUID(), indexId.getName(), shardId);
-        if (evictedShards.add(shardEviction)) {
-            threadPool.generic().submit(new AbstractRunnable() {
-                @Override
-                protected void doRun() {
-                    runIfShardMarkedAsEvictedInCache(shardEviction, () -> {
-                        assert shardsEvictionLock.isHeldByCurrentThread(shardEviction);
-                        final Map<CacheKey, CacheFile> cacheFilesToEvict = new HashMap<>();
-                        cache.forEach((cacheKey, cacheFile) -> {
-                            if (shardEviction.matches(cacheKey)) {
-                                cacheFilesToEvict.put(cacheKey, cacheFile);
-                            }
-                        });
-                        for (Map.Entry<CacheKey, CacheFile> cacheFile : cacheFilesToEvict.entrySet()) {
-                            try {
-                                cache.invalidate(cacheFile.getKey(), cacheFile.getValue());
-                            } catch (RuntimeException e) {
-                                assert false : e;
-                                logger.warn(() -> new ParameterizedMessage("failed to evict cache file {}", cacheFile.getKey()), e);
-                            }
+    public void markShardAsEvictedInCache(String snapshotUUID, String snapshotIndexName, ShardId shardId) {
+        synchronized (shardsEvictionsMutex) {
+            if (allowShardsEvictions) {
+                final ShardEviction shardEviction = new ShardEviction(snapshotUUID, snapshotIndexName, shardId);
+                if (addPendingShardEviction(shardEviction)) {
```

Comment:
I agree this simplifies the code. One of the motivation to differentiate pending and running shard evictions was to allow a searchable snapshot shard to immediately execute the eviction of cache files, while in your suggestion it would mean that the shard will have to wait for the submitted runnable to complete in the generic thread pool before moving forward with recovery. Maybe we are OK with this?

Reply:
That is a good point, thanks for clarifying.

With the current threading this seems impossible to happen, since `evictCacheFilesIfNeeded` is [always called on the generic thread pool](https://github.com/elastic/elasticsearch/blob/672972cc55537fe92d382f5f0bc65696c5a35a48/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotIndexEventListener.java#L62). Since the evict of shard folders must happen before restoring the same searchable snapshot shard again, the initial shard evict job would come before a subsequent restore of the same searchable snapshot shard in the generic queue.

I would be inclined to rely on this for now. If those were separate threads, I suppose waiting would like be out of the question so would require a new design if we change the threading model.

Reply:
>  the initial shard evict job would come before a subsequent restore of the same searchable snapshot shard in the generic queue

Thanks Henning, I think you're right. I went with your suggestion of simplification.